src/transaction/log_replication.cpp

```diff
@@ -353,17 +353,21 @@ namespace cublog
 
   template <typename T>
   void
-  replicator::read_and_redo_record (cubthread::entry &thread_entry, LOG_RECTYPE rectype,
-				    const log_lsa &prev_rec_lsa, const log_lsa &rec_lsa)
+  replicator::read_and_redo_record (cubthread::entry &thread_entry, const LOG_RECORD_HEADER &rec_header,
+				    const log_lsa &rec_lsa)
   {
     m_redo_context.m_reader.advance_when_does_not_fit (sizeof (T));
-    const log_rv_redo_rec_info<T> record_info (rec_lsa, rectype,
+    const log_rv_redo_rec_info<T> record_info (rec_lsa, rec_header.type,
 	m_redo_context.m_reader.reinterpret_copy_and_add_align<T> ());
 
     // only mvccids that pertain to redo's are processed here
     const MVCCID mvccid = log_rv_get_log_rec_mvccid (record_info.m_logrec);
     assert_correct_mvccid (record_info.m_logrec, mvccid);
-    log_replication_update_header_mvcc_vacuum_info (mvccid, prev_rec_lsa, rec_lsa, m_bookkeep_mvcc_vacuum_info);
+    log_replication_update_header_mvcc_vacuum_info (mvccid, rec_header.back_lsa, rec_lsa, m_bookkeep_mvcc_vacuum_info);
+    if (m_replicate_mvcc && MVCCID_IS_NORMAL (mvccid))
```

Comment:
`MVCC_IS_NORMAL` or `MVCCID_IS_VALID` ?
I suppose `MVCCID_IS_VALID` is more correct.

Reply:
MVCCID_IS_VALID contains the special value MVCCID_ALL_VISIBLE (I think is used to replace an MVCCID when is not possible to change the record size - to remove the MVCCID, but we still want to set the data record as visible).
I think here, we don't deal with data records, but with log records and we don't expect to have the value MVCCID_ALL_VISIBLE.
Any reason not to have an assert instead of if (.. MVCCID_IS_NORMAL (mvccid)) ?

Reply:
The function `read_and_redo_record` is typed for both non-mvcc and mvcc-containing log records (`LOG_REDO_DATA`, `LOG_MVCC_REDO_DATA`, `LOG_UNDOREDO_DATA`, ..). Thus, the function `log_rv_get_log_rec_mvccid` is also a template function which will, depending on implementation return either `MVCCID_NULL` or a valid one from the log record. That's why we need to have the check.